File: nanobot-security-layer/nanobot_security_layer/loader.py
from __future__ import annotations

import json
import logging
import os
from pathlib import Path
from typing import TYPE_CHECKING

from .plugins.policy_gate import PolicyGatePlugin
from .types import SecurityConfig, PluginEntry

if TYPE_CHECKING:
    from .types import SecurityPlugin

logger = logging.getLogger("nanobot_security_layer")

DEFAULT_CONFIG_PATHS = [
    "/opt/nanobot/config/security-policy.json",
    Path.home() / ".nanobot" / "config" / "security-policy.json",
]

BUILTIN_FACTORIES: dict[str, type] = {
    "policy-gate": PolicyGatePlugin,
}


def load_security_config() -> SecurityConfig:
    config_path = os.environ.get("SECURITY_POLICY_PATH")
    if not config_path:
        for candidate in DEFAULT_CONFIG_PATHS:
            if Path(candidate).is_file():
                config_path = str(candidate)
                break

    if not config_path or not Path(config_path).is_file():
        logger.warning("No security config found, pipeline will have no plugins")
        return SecurityConfig(plugins=[])

    try:
        raw = Path(config_path).read_text(encoding="utf-8")
        data = json.loads(raw)
        entries = [
            PluginEntry(
                id=e["id"],
                enabled=e.get("enabled", True),
                priority=e.get("priority", 100),
                config=e.get("config"),
            )
            for e in data.get("plugins", [])
        ]
        logger.info("Loaded config from %s (%d plugin entries)", config_path, len(entries))
        return SecurityConfig(plugins=entries)
    except Exception:
        logger.exception("Failed to parse security config at %s", config_path)
        return SecurityConfig(plugins=[])


async def create_plugins(config: SecurityConfig) -> list[SecurityPlugin]:
    plugins: list[SecurityPlugin] = []

    for entry in config.plugins:
        if not entry.enabled:
            continue

        factory_cls = BUILTIN_FACTORIES.get(entry.id)
        if not factory_cls:
            logger.warning("Unknown plugin '%s', skipping", entry.id)
            continue

        try:
            plugin = factory_cls()
            plugin.priority = entry.priority
            await plugin.initialize(entry.config or {})
            plugins.append(plugin)
            logger.info("Plugin '%s' initialized (priority=%d)", entry.id, entry.priority)
        except Exception:
            logger.exception("Failed to initialize plugin '%s'", entry.id)

    return plugins

File: openclaw-security-layer/src/loader.ts
import * as fs from "node:fs";
import * as path from "node:path";
import type { SecurityConfig, SecurityPlugin, PluginEntry } from "./types.js";
import { PolicyGatePlugin } from "./plugins/policy-gate.js";

const DEFAULT_CONFIG_PATHS = [
  "/root/.openclaw/config/security-policy.json",
  path.join(process.env.HOME ?? "", ".openclaw/config/security-policy.json"),
];

const BUILTIN_FACTORIES: Record<string, () => SecurityPlugin> = {
  "policy-gate": () => new PolicyGatePlugin(),
};

export function loadSecurityConfig(): SecurityConfig {
  const configPath = process.env.SECURITY_POLICY_PATH
    ?? DEFAULT_CONFIG_PATHS.find((p) => fs.existsSync(p));

  if (!configPath || !fs.existsSync(configPath)) {
    console.error("[SecurityLayer] No security config found, pipeline will have no plugins");
    return { plugins: [] };
  }

  try {
    const raw = fs.readFileSync(configPath, "utf-8");
    const config = JSON.parse(raw) as SecurityConfig;
    console.error(`[SecurityLayer] Loaded config from ${configPath} (${config.plugins.length} plugin entries)`);
    return config;
  } catch (err) {
    console.error("[SecurityLayer] Failed to parse security config:", err);
    return { plugins: [] };
  }
}

export async function createPlugins(config: SecurityConfig): Promise<SecurityPlugin[]> {
  const plugins: SecurityPlugin[] = [];

  for (const entry of config.plugins) {
    if (!entry.enabled) continue;

    const factory = BUILTIN_FACTORIES[entry.id];
    if (!factory) {
      console.error(`[SecurityLayer] Unknown plugin '${entry.id}', skipping`);
      continue;
    }

    try {
      const plugin = factory();
      plugin.priority = entry.priority;
      await plugin.initialize(entry.config ?? {});
      plugins.push(plugin);
      console.error(`[SecurityLayer] Plugin '${entry.id}' initialized (priority=${entry.priority})`);
    } catch (err) {
      console.error(`[SecurityLayer] Failed to initialize plugin '${entry.id}':`, err);
    }
  }

  return plugins;
}
